sagan/models/attention.py

```diff
@@ -1,39 +1,56 @@
 import tensorflow as tf
-from tensorflow import keras
-from tensorflow.keras import layers, initializers
+from tensorflow.keras import layers
+
+from sagan.models.spectral_norm import SpectralNorm
 
 
 class SelfAttention(layers.Layer):
-    def __init__(
-        self, input_dims, trainable=True, name=None, dtype=None, dynamic=False, **kwargs
-    ):
-        super().__init__(
-            trainable=trainable, name=name, dtype=dtype, dynamic=dynamic, **kwargs
-        )
-        self.query_convolution = layers.Conv2D(filters=input_dims // 8, kernel_size=1)
-        self.key_convolution = layers.Conv2D(filters=input_dims // 8, kernel_size=1)
-        self.value_convolution = layers.Conv2D(filters=input_dims // 8, kernel_size=1)
+    def __init__(self):
+        super(SelfAttention, self).__init__()
 
     def build(self, input_shape):
-        self.gamma = self.add_weight(
-            self.name + "_gamma", shape=(), initializer=initializers.Zeros()
+        n, h, w, c = input_shape
+        self.n_feats = h * w
+        self.conv_theta = layers.Conv2D(
+            c // 8,
+            1,
+            padding="same",
+            kernel_constraint=SpectralNorm(),
+            name="Conv_Theta",
         )
-
-    def call(self, inputs, *args, **kwargs):
-        query = self.query_convolution(inputs)
-        key = self.key_convolution(inputs)
-        value = self.value_convolution(inputs)
-        query_dim = tf.shape(query)
-        batch_size, height, width = query_dim[0], query_dim[1], query_dim[2]
-        proj_query = tf.reshape(query, (batch_size, height * width, -1))
-        proj_key = tf.transpose(
-            tf.reshape(key, (batch_size, height * width, -1)), (0, 2, 1)
+        self.conv_phi = layers.Conv2D(
+            c // 8, 1, padding="same", kernel_constraint=SpectralNorm(), name="Conv_Phi"
         )
-        proj_value = tf.transpose(
-            tf.reshape(value, (batch_size, height * width, -1)), (0, 2, 1)
+        self.conv_g = layers.Conv2D(
```

Comment:
@sayannath I think its better if we follow a naming convention of `f(x)`, `g(x)` and `h(x)` for the self-attention block, or even better, we could go with key, query and value in order for the code to be even more readable.

Reply:
Cool! will incorporate these changes.